explore/marginal-gain/1.Marginal-Gain-Multiple-Mutations.py

```diff
@@ -0,0 +1,253 @@
+    '7428': 'VHL',    # von Hippel-Lindau tumor suppressor
+    '29126': 'CD274', # CD274 molecule
+    '672': 'BRCA1',   # BRCA1, DNA repair associated
+    '675': 'BRCA2',   # BRCA2, DNA repair associated
+    '238': 'ALK',     # anaplastic lymphoma receptor tyrosine kinase
+    '4221': 'MEN1',   # menin 1
+    '5979': 'RET',    # ret proto-oncogene
+}
+
+
+# ## Median absolute deviation feature selection
+
+# In[81]:
+
+def fs_mad(x, y):
+    """    
+    Get the median absolute deviation (MAD) for each column of x
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+# ## Define pipeline and Cross validation model fitting
+
+# In[82]:
+
+# Parameter Sweep for Hyperparameters
+
+param_grid = {
+    'classify__loss': ['log'],
+    'classify__penalty': ['elasticnet'],
+    'classify__alpha': [1],#10.0 ** np.linspace(-3, 1, 10),
+    'classify__l1_ratio': [0.15],
+}
+
+expression_feats = Pipeline(steps=[
+    ('dim_red', FunctionTransformer(lambda X: X[:,covariates.shape[1]:])),
+    ('select', SelectKBest(fs_mad,2000)),
+])
+
+covariate_feats = Pipeline(steps=[
+    ('dim_red', FunctionTransformer(lambda X: X[:,:covariates.shape[1]])),
+])
+
+combo_pipeline = Pipeline([
+    ('imputer', Imputer()),
+    ('standardize', StandardScaler()),
+    ('features', FeatureUnion([
+        ('covariates_feats', covariate_feats),           
+        ('expression_feats', expression_feats)
+    ])),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+covariates_pipeline = Pipeline(steps=[
+    ('imputer', Imputer()),
+    ('standardize', StandardScaler()),
+    ('select', SelectKBest(fs_mad,'all')),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+cv_pipeline = {}
+cv_pipeline['model a'] = GridSearchCV(estimator=covariates_pipeline, param_grid=param_grid, scoring='roc_auc')
+cv_pipeline['model b'] = GridSearchCV(estimator=combo_pipeline, param_grid=param_grid, scoring='roc_auc')
+
+
+# ## Functions to get statistics for a given model 
+
+# In[83]:
+
+# Get statistics for a given model. 
+
+def get_aurocs(X, y, pipeline, series, model_type):
+    """
+    Fit the classifier for the given mutation (y) and output predictions for it
+    """
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
+    y_train[0] = 1
+    y_test[0] = 1
+    pipeline.fit(X=X_train, y=y_train)
+    y_pred_train = pipeline.decision_function(X_train)
+    y_pred_test = pipeline.decision_function(X_test)
+    cv_score_df = grid_scores_to_df(pipeline.grid_scores_)
+    coeff_df = get_coeffs(pipeline, X_train, model_type)
+    n_pos, n_neg = get_sign_coeffs(coeff_df)
+    cov_ranks = get_ranks_covariates_feat(coeff_df)
+    series['mean_cv_auroc'] = cv_score_df.score.max()
+    series['training_auroc'] = roc_auc_score(y_train, y_pred_train)
+    series['testing_auroc'] = roc_auc_score(y_test, y_pred_test)
+    series['n_pos_coeffs'] = n_pos
```

Comment:
Gotcha. Updated. As you might expect, the combined model shows the clearest advantage where we have a lot of positive examples to work from (TP53).